src/models/AccessControlUnit.ts

```diff
@@ -0,0 +1,318 @@
+    Allowed,
+}
+
+/**
+ * A description of the access an entity has.
+ * If the access is `Banned`, then a single rule that bans the entity will be included.
+ */
+export interface EntityAccess {
+    readonly outcome: Access,
+    readonly rule?: ListRule,
+}
+
+/**
+ * This allows us to work the access an entity has to some thing based on a set of watched/unwatched lists.
+ */
+export default class AccessControlUnit {
+    private readonly userBans = new ListRuleCache(RULE_USER, Recommendation.Ban);
+    private readonly serverBans = new ListRuleCache(RULE_SERVER, Recommendation.Ban);
+    private readonly userAllows = new ListRuleCache(RULE_USER, Recommendation.Allow);
+    private readonly serverAllows = new ListRuleCache(RULE_SERVER, Recommendation.Allow);
+    private readonly caches = [this.userBans, this.serverBans, this.userAllows, this.serverAllows]
+
+    constructor(policyLists: PolicyList[]) {
+        policyLists.forEach(this.watchList, this);
+    }
+
+    public watchList(list: PolicyList) {
+        for (const cache of this.caches) {
+            cache.watchList(list);
+        }
+    }
+
+    public unwatchList(list: PolicyList) {
+        for (const cache of this.caches) {
+            cache.watchList(list);
+        }
+    }
+
+    /**
+     * Test whether the server is allowed by the ACL unit.
+     * @param domain The server name to test.
+     * @returns A description of the access that the server has.
+     */
+    public testServer(domain: string): EntityAccess {
+        return this.testEntity(domain, this.serverAllows, this.serverBans);
+    }
+
+    /**
+     * Test whether the user is allowed by the ACL unit.
+     * Does not test the domain of the user id.
+     * @param mxid The user id to test.
+     * @returns A description of the access that the user has.
+     */
+    public testUserWithoutServer(mxid: string): EntityAccess {
+        return this.testEntity(mxid, this.userAllows, this.userBans);
+    }
+
+    /**
+     * Test whether the user is allowed by the ACL unit. Does take the user's server into consideration.
+     * @param mxid The user id to test.
+     * @returns A description of the access that the user or their server has.
+     */
+    public testUser(mxid: UserID): EntityAccess {
+        const userAccess = this.testUserWithoutServer(mxid.toString());
+        if (userAccess.outcome === Access.Allowed) {
+            return this.testServer(mxid.domain);
+        } else {
+            return userAccess;
+        }
+    }
+
+    private testEntity(entity: string, allowCache: ListRuleCache, bannedCache: ListRuleCache): EntityAccess {
+        // Check if the entity is explicitly allowed.
+        // We have to infer that a rule exists for '*' if the allowCache is empty, otherwise you brick the ACL.
+        const allowRule = allowCache.test(entity);
+        if (allowRule === null && !allowCache.isEmpty()) {
+            return { outcome: Access.NotAllowed }
```

Comment:
It could, but it's irrelevant since they aren't on the allow to begin with. The semantics follow the same algorithm as https://spec.matrix.org/v1.4/client-server-api/#mroomserver_acl